MASSIF-1 Crystal Direct Harvester implementation

jbflo opened the pull request:
This code related to the implementation of the Crystal Direct Harvester in Massif-1 ESRF

This should not have breaking change. 
Beamline who does not have an Harvester should not have any new things/modification to do 
A PR in Mxcube-web related to this one need to be reviewed as well .

jbflo committed:
Remove unnecessary check

Plate is like an SC

jbflo committed:
Add Harvester Hardware Object  and Mockup

jbflo committed:
Add crystalUUID in queue model object

jbflo committed:
Update to fit the new Crims api url

Commit:
Add Get Room temperature command

Commit:
Add function to send info to Crims

Commit:
Add pin cleaning boolean param when Load from harvester

Commit:
Add Custom centering for Harvester

Commit:
Specific logic to load from Harvester

Commit:
file Renamed

Commit:
load a pin from Harvester for calibration

Add possibility to set Temperature mode and

Commit:
Use Object name  instead

Not only Diffractometer use exporter command

Commit:
Custom phase change  should not be when plate mode

Commit:
use Return instead of raise

Temporary  return False for any  Robot Exception, Except Pin Cleaning Station Exception

Commit:
Remove comment and unused function

leave some comment for later

Commit:
Rename  method and add more doc-string

Commit:
Add _mount_from_harvester abstract method

Not use if I should i it here. but that will prevent key error in other  SC Class

fabcor-maxiv commented on mxcubecore/Command/exporter/ExporterClient.py:
```diff
@@ -127,7 +127,7 @@ def __process_return(self, ret):
         """
         if ret[:4] == RET_ERR:
-            msg = "Diffractometer: {}".format(str(ret[4:]))
+            msg = f"{self.get_server_object_name()} : {str(ret[4:])}"
             logging.getLogger("HWR").error(msg)
```
As far as I know, the recommendation for Python logging is to always use printf style formatting (the one with the `%` interpolation) to take advantage of lazy evaluation:

* https://realpython.com/python-f-strings/#lazy-evaluation-in-logging
* https://docs.python.org/3/library/stdtypes.html#old-string-formatting

fabcor-maxiv commented on mxcubecore/HardwareObjects/Beamline.py:
```diff
@@ -414,18 +414,42 @@ def sample_changer_maintenance(self):
+        """Harvester Hardware object
+        can be a sample or plate holder
+
+        Returns:
+            Optional[AbstractHarvester]:
```
Why not use type hints?

jbflo replied:
As the other method are like this, i'll leave it  for now. maybe someone can make the change for all the methods in this class in a separate PR.

rhfogh replied:
Ah well, in that case it should stay where it is. I do not seem to have a very high hit-rate today )-:

fabcor-maxiv reviewed (commented):
A couple of random comments. I did not review the whole thing, we do not seem to have such harvester.

Commit:
Remove commented class, add deprecated flag instead

Commit:
Revert formatting style

beteva commented on mxcubecore/HardwareObjects/abstract/AbstractSampleChanger.py:
```diff
@@ -715,6 +715,10 @@ def _do_unload(self, sample_slot=None):
     def _do_reset(self):
         return
+    
+    @abc.abstractmethod
```
I would remove the abstract decorator, as it is only specific for the harvester and should not require that all the sample changers implement this method.

Commit:
remove the abstract decorator

Commit:
Flake/ Black reformatted

Add default timeout 0

beteva commented on mxcubecore/Command/exporter/ExporterClient.py:
```diff
@@ -127,7 +127,7 @@ def __process_return(self, ret):
         """
         if ret[:4] == RET_ERR:
-            msg = "Diffractometer: {}".format(str(ret[4:]))
+            msg = "{} : {}".format(self.get_server_object_name(), str(ret[4:]))
```
You might as well consider use the f-string (see with pylint).

jbflo replied:
I did, but then followed Fabcor suggestion 
https://github.com/mxcube/mxcubecore/pull/826#discussion_r1417609839

beteva commented on mxcubecore/HardwareObjects/Beamline.py:
```diff
@@ -458,10 +456,36 @@ def sample_changer_maintenance(self):
 
     __content_roles.append("sample_changer_maintenance")
 
+    @property
+    def harvester(self):
+        """Harvester Hardware object
+        can be a sample or plate holder
+
+        Returns:
+            Optional[AbstractHarvester]:
```
Is this needed? It can be just seen as a sample changer in beamline.

jbflo replied:
Yes, yes but it is a bit more complex.
- This Hardware can not be treated as a sample changer as it is more like a sample Holder. 
-Accessing it is possible through the Sample Changer but In Massif-1 they still wanted to use the Harvester even the Sample Changer is not available.

beteva replied:
My question was more can't you specify the harvester hardware object instead of the sample changer in the beamline configuration. Thus you can also benefit from the checks which are done for mounting/unmounting.

jbflo replied:
> Is this needed? It can be just seen as a sample changer in beamline.

@beteva  As we discuss a while ago, could you have look again ?

beteva commented on mxcubecore/HardwareObjects/Beamline.py:
```diff
@@ -458,10 +456,36 @@ def sample_changer_maintenance(self):
+
+        Returns:
+            Optional[AbstractHarvester]:
+        """
+        return self._objects.get("harvester")
+
+    __content_roles.append("harvester")
+
+    @property
+    def harvester_maintenance(self):
+        """harvester maintenance Hardware object
+
+        Returns:
+            Optional[Harvester]:
+        """
+        return self._objects.get("harvester_maintenance")
+
+    __content_roles.append("harvester_maintenance")
```
Is this needed? It can be just seen as a sample changer in beamline.

jbflo replied:
https://github.com/mxcube/mxcubecore/pull/826#discussion_r1449178378

beteva commented on mxcubecore/HardwareObjects/EMBLFlexHCD.py:
```diff
@@ -183,31 +189,58 @@ def init(self):
 
+    def _mount_from_harvester(self):
+        return self._harvester and self._harvester_hwo
+
```
@jbflo, please see my comment for the AbstractSampleChanger.

jbflo replied:
method removed from AbstractSampleChanger

beteva commented on mxcubecore/HardwareObjects/EMBLFlexHCD.py:
```diff
@@ -183,31 +189,58 @@ def init(self):
+                    ):
+                        present_sample_list.append(sample)
 
         return present_sample_list
```
For better readability, it would be good to split this in three methods - the get_sample_list, which is the standard method, which calls  _get sample_list_flex and _get_sample_list_harvester.

beteva commented on mxcubecore/HardwareObjects/EMBLFlexHCD.py:
```diff
@@ -531,25 +608,38 @@ def _do_load(self, sample=None):
             for msg in self.get_robot_exceptions():
                 logging.getLogger("user_level_log").error(msg)
-            raise
+            return False
```
Attention, this is changing the behaviour of the method.

jbflo replied:
Yes, thanks  We have to see on Other Beamlines, but in Massif-1 this behavoir is OK in since
Just raise will return None which consider a loading success.
But to avoid  to collect on the previous mounted sample.  a return False fit better.

rhfogh replied:
I would not agree here, J-B. The behaviour of load (which is determined by _do_load) must be the same as the behaviour of SampleChanger.load(). Replacing a raised error with a return of False does not look like the same behaviour.

Also - if MASSIF-1 really interprets a call that raises an error as a successful load, that sounds like a serious error that should eb fixed on the MASSIF-1 side, no?

jbflo replied:
not only   MASSIF-1,  I mentioned Massif-1 because it's there I found the issue. 
this code bellow is not quite OK  ? why if load return None , you make it True ?
https://github.com/mxcube/mxcubeweb/blob/a958e7d1d3e4ffcc354974d07685522088f7cd23/mxcubeweb/core/components/samplechanger.py#L198C1-L205C31

rhfogh replied:
I'd say the problem is that happens if load raises an error. The way this would normally work in Python, the raised error would stop further execution of the code and jump right down to line 224 where it says:

        except Exception as ex:
            logging.getLogger("MX3.HWR").exception("[SC] sample could not be mounted")
            raise RuntimeError(str(ex)) from ex

Since this again raises an error, you would never get to line 244 
        return res
but would have the error propagate upwards to the next error handler. 
With your change you would get a different behaviour: you would have the function return False instead of raising an error. If you are making that change I would say you need  a very careful analysis of the calling functions and their error handling to be sure that you want the new behaviour rather than the old one.

beteva commented on mxcubecore/HardwareObjects/Harvester.py:
```diff
@@ -0,0 +1,501 @@
+        <exporter_address>wid30harvest:9001</exporter_address>
+    </equipment>
+-----------------------------------------------------------------
+"""
```
It is great that there is some explanation, Could you, please, add the standard MXCuBE license message (see EMBLFlexHCD.py) Also the xml example should use object instead of equipment.

beteva commented on mxcubecore/HardwareObjects/HarvesterMaintenance.py:
```diff
@@ -0,0 +1,140 @@
+"""
+Harvester Maintenance maintenance .
+"""
```
Same as for Harvester.

beteva commented on mxcubecore/HardwareObjects/abstract/AbstractSampleChanger.py:
```diff
@@ -679,6 +678,13 @@ def _resolve_component(self, component):
             return comp
         return component
 
+    def _mount_from_harvester(self):
+        """
+        Mount from Harvester or from default dewar
+        Returns: (Bool) true Mount from Harvester
+        """
+        return
+
```
Is this used to check if harvester or sample changer is used? If yes, than the name is misleading (maybe something like is_harvester is more appropriated) and moreover it should be a property.

Commit:
rename of method

Commit:
Add Doc-string and Type-hint format 

 Split method and fix flake8 warnning

Commit:
add MXCuBE license message

rhfogh reviewed (changes requested):
Surely a file called EMBLFlexHCD.py should be in HardwareObjects/EMBL, and not in HardwareObjects/ ?

beteva commented:
Not quite @rhfogh, in this case it refers that the flex is manifactured by EMBL, just as Cats, Marvin ...:-)

Commit:
Use f string again

Commit:
Remove abstract function /  Fix Typo

Commit:
Expose rest token

Commit:
Improve whether the sample has protein acronym or not

jbflo commented:
> Surely a file called EMBLFlexHCD.py should be in HardwareObjects/EMBL, and not in HardwareObjects/ ?

@rhfogh  do you still request this change  ?

Commit:
Autoflake Failed > Now Passed

crystal_symmetry and yaml where imported twice

marcus-oscarsson commented:
Looks good as far as I can see :), (its a big one ;) )

beteva commented on mxcubecore/HardwareObjects/abstract/sample_changer/Crims.py:
```diff
@@ -78,15 +75,23 @@ def __init__(self, *args):
-        xml = f.read()
+
+        headers = { 
+            'User-Agent' : "Mozilla/5.0 (Windows NT 6.1; Win64; x64",
```
Should this really be hard-coded?

beteva commented:
Hi @jbflo, Most of the changes are Ok with me. It still bothers me, though, that you have to check for harvester in the EMBLFlexHCD.py.

jbflo commented on mxcubecore/HardwareObjects/Harvester.py:
```diff
@@ -0,0 +1,643 @@
+
+        return None
+
+    def check_crystal_state(self, crystal_uuid: str) -> str | None:
```
@beteva @marcus-oscarsson I Just Notice  this Type hints
require Python 1.10 and above.
Is this a problem ?

marcus-oscarsson replied:
You mean ` str | None` ?

You should be able to use 

```
from typing import Optional

 ... -> Optional[str]
```

jbflo replied:
> You mean ` str | None` ?
Yes
> You should be able to use
> 
> ```
> from typing import Optional
>  ... -> Optional[str]
> ```
Yeah,

jbflo commented:
> Hi @jbflo, Most of the changes are Ok with me. It still bothers me, though, that you have to check for harvester in the EMBLFlexHCD.py. 

as we We check because depending of where the SC take the Sample , the Sample List and the Next Actions are different 

maybe it's because the Dewar is (consider) a part of the FLEX.

Now we should probably think of 2 things.
1- The Sample changer ( the Robot ) FLEX
2 The Sample Holder (Can be a Dewar(HCD) or Harvester)

Or maybe we should have a Flex and a HCD Class ?  Or an EMBLFlexHavester CLass ?
having them both implementing as one is not much help understanding the Harvester.

Yeah maybe an EMBLFlexHavester CLass but this will have must of the EMBLFlexHCD methode In.

I am thinking about it @beteva , but for now i don't have a better clue.
for now I Think this class (EMBLFlexHCD) is the best place. 

please let me know if you are thinking  of a better logic.

Commit:
Lint and Python 3.9 compatibility

Commit:
Remove hard coded User Agent

fabcor-maxiv commented on mxcubecore/HardwareObjects/abstract/sample_changer/Crims.py:
```diff
@@ -122,5 +129,38 @@ def get_processing_plan(barcode, crims_url, harvester_key):
+def send_data_collection_info_to_crims(crims_url, crystaluuid, datacollectiongroupid, dcid, proposal, rest_token):
+    try:
+        url = crims_url
+        url = (
+            crims_url
+            + str(crystaluuid)
+            + "/dcgroupid/"
+            + str(datacollectiongroupid)
+            + "/dcid/"
+            + str(dcid)
+            + "/mx/"
+            + str(proposal)
+            + "/token/"
+            + str(rest_token)
+            + "?janitor_key=kbonvRqc8"
```
Maybe better as an f-string?

Also is the key supposed to be here? Is that a secret?

Commit:
Converte to f string

Remove hard coded key

Commit:
[Harvester] Fix Url parameter

Add Test  data to Mock up / Fix Type Hint

marcus-oscarsson commented:
I agree with @beteva that it would be nice if the sample changer and the Harvester code could be de-coupled. I think that would be a nice followup PR.

beteva commented:
> I am thinking about it @beteva , but for now i don't have a better clue. for now I Think this class (EMBLFlexHCD) is the best place.
> 
> please let me know if you are thinking of a better logic.

@jbflo, it might be tempting to split Robot and Holder, but I am afraid that this will add an extra complication to already complicated device. I would rather go for the other idea - having a specific  EMBLFlexHavester CLass, which inherits from the EMBLFlexHCD and only overloads the methods which differ for the harvester. Than in the beamlise-setup.yml it will be easy to choose which "sample changer" you have at given moment.What do you think?

jbflo commented:
> > I am thinking about it @beteva , but for now i don't have a better clue. for now I Think this class (EMBLFlexHCD) is the best place.
> > please let me know if you are thinking of a better logic.
>
> @jbflo, it might be tempting to split Robot and Holder, but I am afraid that this will add an extra complication to already complicated device. I would rather go for the other idea - having a specific  EMBLFlexHavester CLass, which inherits from the EMBLFlexHCD and only overloads the methods which differ for the harvester. Than in the beamlise-setup.yml it will be easy to choose which "sample changer" you have at given moment.What do you think?

Yes I also think it's a better Idea , I'll go on that side. Thanks !

Commit:
[Harvester] Remove Harvester Logic in  EMBLFlexHCD class

Commit:
[Harvester] add specific EMBLFlexHavester CLass

 which inherits from the EMBLFlexHCD

beteva commented on mxcubecore/HardwareObjects/EMBLFlexHarvester.py:
```diff
@@ -0,0 +1,207 @@
+class EMBLFlexHarvester(EMBLFlexHCD):
+    __TYPE__ = "Flex Sample Changer"
+
+    def __init__(self, *args, **kwargs):
+        super(EMBLFlexHarvester, self).__init__(self.__TYPE__, True, *args, **kwargs)
```
You can do
`super().__init__(*args, **kwargs)` 
Also self._loaded_sample = (-1, -1, -1) is better to be here, than in init(). you can also add self._harvester = None (not compulsory, just style).

beteva commented on mxcubecore/HardwareObjects/EMBLFlexHarvester.py:
```diff
@@ -0,0 +1,207 @@
+            + "%02d" % loaded_sample[2]
+        )
+
+    def get_loaded_sample(self):
+        sample = None
+
+        loaded_sample_addr = self._hw_get_mounted_sample()
+
+        for s in self.get_sample_list():
+            if s.get_address() == loaded_sample_addr:
+                sample = s
+
+        return sample
```
get_loaded_sample() is identical to the parent class. no need to overload it

beteva replied:
Please, check the style with black as well.

beteva reviewed (approved):
Well done @jbflo. All looks nice and neat. I only have few style remarks. Have not checked the EMBLFlexHCD code, but I trust you it is the original one.

Commit:
[Harvester] Black & remove unused Property

Remove  Duplicated line

marcus-oscarsson commented on mxcubecore/HardwareObjects/EMBLFlexHCD.py:
```diff
@@ -521,6 +521,16 @@ def defreeze(self):
         self.enable_power()
         self._execute_cmd_exporter("defreezeGripper", command=True)
 
+    def get_room_temperature_mode(self):
```
I guess these two methods should not be here then, if we now have the `EMBLFlexHarvester` object

Commit:
Keep EMBLFlexHCD As it was / Just keep the lint  fix

Commit:
[Flex] Add possibility to set and get Room temperature Mode

marcus-oscarsson commented on mxcubecore/HardwareObjects/Beamline.py:
```diff
@@ -774,9 +798,9 @@ def get_default_acquisition_parameters(self, acquisition_type="default"):
-        #""")
+        # logging.getLogger("HWR").debug(f"""
+        # Beamline object. Getting acquisition parameters for acquisition type {acquisition_type}
+        # """)
```
Remove or un-comment ?

jbflo replied:
I could but i Prefer not making the remove part of this PR. 

it Was already commented.
by @ Andrey Gruzinov Maybe ?
I just applied black on it so it could pass the Linter.

marcus-oscarsson replied:
I see, black is good but Id say commented code like this can simply be removed ;)

jbflo replied:
comment removed

marcus-oscarsson commented on mxcubecore/HardwareObjects/EMBLFlexHCD.py:
```diff
@@ -378,7 +378,7 @@ def chained_load(self, old_sample, sample):
     def _set_loaded_sample_and_prepare(self, sample, previous_sample):
         res = False
 
-        if not -1 in sample and sample != previous_sample:
```
Looks good :) :+1:

marcus-oscarsson commented on mxcubecore/HardwareObjects/EMBLFlexHarvester.py:
```diff
@@ -0,0 +1,207 @@
+# encoding: utf-8
```
Looks good as far as I can see

marcus-oscarsson commented on mxcubecore/HardwareObjects/EMBLFlexHarvester.py:
```diff
@@ -0,0 +1,207 @@
+        )
+        return self.get_room_temperature_mode()
+
+    def mount_from_harvester(self):
```
Does this mean that there should be an equivalent method for the AbstractSampleChanger that returns false by default, or how is this used ?

jbflo replied:
HAHA :smiley: , Having an AbstractSampleChanger  equivalent method that returns false by default for 
mount_from_harvester is  indeed what I wanted.

but could not agree on this with @beteva 
Or maybe my explanation was not clear enough 
SO I find a work around and ending up removing it from AbstractSampleChanger.

marcus-oscarsson replied:
But how then are you using this method of the base class does not have this method any other sample changer would throw an `AttributeError`

beteva replied:
I still do not like to have anything about harvester in the AbstractSampleChanger class. For me, unless I do not get it right, the _do_load method should be overloaded with this one.

jbflo replied:
> But how then are you using this method of the base class does not have this method any other sample changer would throw an `AttributeError`

`    
       sc = HWR.beamline.sample_changer

        try:
            mount_from_harvester = sc.mount_from_harvester()
        except AttributeError:
            mount_from_harvester = False
`
            
I did this way every time.  not elegant but  SC who does not this method should pass the AttributeError

marcus-oscarsson replied:
But, why do you need to call `mount_from_harvester` specifically would it not be enough to call `mount` and the `mount` of the `Harvester` object knows what to do ?

jbflo replied:
There are a lot of mounting logic there : mxcubeweb/core/components/samplechanger.py  

And mount_from_harvester help us prevent or add  behavior that not or are specific when mounting from an Harvester.  
 
have a look at this PR: [https://github.com/mxcube/mxcubeweb/pull/1145/files](url)
FIle :  mxcubeweb/core/components/samplechanger.py  


PS : those two PR are a big big (Sorry about that), but reviewing them both together maybe better

marcus-oscarsson replied:
I see, how about simply setting a flag on the harvester hardware object and deal with the mounting specifics in the harvester hardware object. I assume that you know precisely when you need to use harvester. You could  add something to the sample address so that you know that its a harvester sample. In this case the harvester hardware object would know if its a standard sample or a harvester sample.

jbflo replied:
That will  unfortunately not enough :  mxcubeweb.mxcubeweb.core.components.samplechanger.py
has specific logic related to queuing, loading and  centering that are by default.

Things we don't want : default autoloop centring
Things we want do to : add logic while execution of queue 

it's not just mounting, have another look at  this file : [https://github.com/mxcube/mxcubeweb/blob/JB-develop-harvester/mxcubeweb/core/components/samplechanger.py](url)
check on method like queue_mount_sample or mount_sample_clean_up.

marcus-oscarsson commented on mxcubecore/model/queue_model_objects.py:
```diff
@@ -329,6 +329,14 @@ def has_lims_data(self):
             return False
         return False
 
+    def has_protein_acronym(self):
```
Can't we somehow merge this with has_lims_data so that we dont need to check for both here :https://github.com/mxcube/mxcubecore/pull/826/files#diff-bf4145a2e91e5b6388dbc69591ea8c2fc19b54fcdf10087cd77b0087ccb3b423R283

jbflo replied:
Hmm I think not . 
unfortunately  has_lims_data logic  is  not suitable for all cases. 
in this case protein acronym is known even there is no lims data .

And The scientist would ask for it again if we remove the second check.
Having the acronym as parts in the prefix was a mandatory for them.

we could maybe merge  has_protein_acronym with has_lims_data ?
 
 
 PS : thanks i should not use TypeError here i'll fix it,

marcus-oscarsson replied:
> we could maybe merge has_protein_acronym with has_lims_data ?

yes, it that would be nice if we could try that

jbflo replied:
@marcus-oscarsson I had another look and i think a merge is not a good fit for this logic.

has_lims_data method name is too generic would have been better as has_lims_id ..
has_protein_acronym is a better fit on get_default_prefix as you only wanted to care about the protein_acronym


`  i would reccommend something like that : 

      if sample_data_node:
            name = sample_data_node.name
            if (sample_data_node.has_protein_acronym()):
                protein_acronym = sample_data_node.crystals[0].protein_acronym
                if name:
                    prefix = "%s-%s" % (protein_acronym, name)
                else:
                    prefix = protein_acronym
            else:
                prefix = name or ""
      elif generic_name:
            prefix = "<acronym>-<name>"
`

and  not check for has_lims_data()

marcus-oscarsson commented on mxcubecore/HardwareObjects/Microdiff.py:
```diff
@@ -260,6 +269,8 @@ def init(self):
         self.centringVertical = self.get_object_by_role("centringVertical")
         self.centringFocus = self.get_object_by_role("centringFocus")
 
+        self.saved_motor_position = {}
```
How is `saved_motor_position` used ?

jbflo replied:
- For calibration purpose (to find find offset position )
 1 -  move those motors to a specific position and saved 
 2 - move again .. and again example(3 click centering)
 3-  then we need the saved position for offsets calculation
 
it being used here : https://github.com/mxcube/mxcubeweb/pull/1145/files#diff-5ea1dcf48d30f15039fd0b4b1bd66d2b46bc68ee3aa35f4f0db10259723ffa79R280

marcus-oscarsson replied:
Hm, are you sure, I could not find any reference to `saved_motor_position`

jbflo replied:
> Hm, are you sure, I could not find any reference to `saved_motor_position`

Yes, Just tagged you , https://github.com/mxcube/mxcubeweb/pull/1145#discussion_r1565904742

rhfogh commented on mxcubecore/HardwareObjects/EMBLFlexHCD.py:
```diff
@@ -546,7 +546,7 @@ def _do_load(self, sample=None):
             self._wait_busy(300)
             logging.getLogger("HWR").info(f"Waited SC activity {time.time() - _tt}")
-        except:
+        except Exception:
```
I think this should actually be 'except:' - whatever the linter says. 
The difference is that 'except:' also catches 'built-in system-exiting exceptions', and in this case you are re-raising the exception anyway and you  would still want to log the robot_exceptions before exiting.

jbflo replied:
hmm should we not care about linter anymore ?
PR should be merge if linter are passed right ?

Not an expert, i'll leave this for the other guys :smiley:

marcus-oscarsson replied:
You can use `BaseException` if you want to catch 'built-in system-exiting exceptions' and still keep the linter happy. However, I don't think we should be catching those, unless there is a good reason for it, even if as in this case we re-raise them.

rhfogh replied:
In this case I think there is a good reason. We want to log those accumulated messages before we exit.
Also, frankly, following the linter in this case is a matter of the tail wagging the dog. That particular linter point was put in to 'encourage' people to put in specific exceptions rather than catching everything willy-nilly. If we want to allow a bare 'except:' in some cases (and I think we do) surely we should either remover that check from the linter or use the mechanisms for making exceptions for a given line of code. Making convoluted python code to be allowed to do what we want while by passing the linter does not sound like the right way to go.

marcus-oscarsson replied:
Well, in this case the sample changer wont raise a system exiting exception and if it does we probably can't ask it to give us the reason for it. If it does we would know what happened by looking at the stacktrace or the sample changer log file.

The second point would be that, its really not a good idea to catch system exiting exceptions because we might trap them and our program might not exit or behave in strange ways. Its problematic when we use CTRL-C or for us more commonly decide to stop the queue by raising `GreenletExit`

rhfogh replied:
OK, I'll take that as a convincing argument. Point withdrawn.

Commit:
Fix Exception  type

Should be AttributeError, IndexError

Commit:
Rename variables.

Removed from UI

Commit:
[more precise logic ] use has_protein_acronym in place of has_lims_data

Apply Spell Check
has_lims_data is used on qt version can't remove but should be mark as Deprecated

Commit:
Lint / Remove Comment

Commit:
Move Function/logic from mxcubeweb to mxcubecore level

Commit:
[FIX] Move Function/logic from mxcubeweb to mxcubecore level
